csrc/dynamic_transform.cpp

```diff
@@ -20,43 +20,253 @@
 
 namespace nvfuser {
 
-//! Gather information about concretizing transformations with
+DynamicTransformInitialInfo DynamicTransformInitialInfo::clone(
+    IrCloner& ir_cloner) const {
+  DynamicTransformInitialInfo cloned_info(
+      static_cast<Fusion*>(ir_cloner.container()));
+  cloned_info.dynamic_reshapes_.reserve(dynamic_reshapes_.size());
+  for (const auto op : dynamic_reshapes_) {
+    if (op) {
+      cloned_info.dynamic_reshapes_.push_back(ir_cloner.clone(op));
+    }
+  }
+  cloned_info.dynamic_resizes_.reserve(dynamic_resizes_.size());
+  for (const auto op : dynamic_resizes_) {
+    if (op) {
+      cloned_info.dynamic_resizes_.push_back(ir_cloner.clone(op));
+    }
+  }
+  cloned_info.expr_eval_ = expr_eval_.clone(ir_cloner);
+  return cloned_info;
+}
+
+std::string DynamicTransformInitialInfo::toString() const {
+  std::stringstream ss;
+  ss << "DynamicTransformInitialInfo\n";
+  std::string indent = "  ";
+  ss << indent << "Dynamic reshapes:\n";
+  for (const auto& op : dynamic_reshapes_) {
+    ss << indent << indent << op->toString() << "\n";
+  }
+  ss << indent << "Dynamic resizes:\n";
+  for (const auto& op : dynamic_resizes_) {
+    ss << indent << indent << op->toString() << "\n";
+  }
+  ss << indent << "Root dynamic Vals:\n";
+  for (const auto& v : root_dynamic_vals_) {
+    ss << indent << indent << v->toString() << "\n";
+  }
+  return ss.str();
+}
+
+//! Gather information about concretizing transformations without
 //! concrete input sizes.
-class DynamicTransformInfoBuilder : public IterVisitor {
+class DynamicTransformInitialInfoBuilder : public IterVisitor {
  public:
-  DynamicTransformInfoBuilder(Fusion* fusion, ExpressionEvaluator* expr_eval);
+  DynamicTransformInitialInfoBuilder(Fusion* fusion) : info_(fusion) {
+    TORCH_INTERNAL_ASSERT(
+        !fusion->isA<kir::Kernel>(),
+        "Invalid container. Kernel container not allowed.\n");
+
+    traverseTo(fusion, fusion->getTerminatingOutputs(), false, false);
+
+    finalizeDynamicVals();
+
+    // initial_info_ provides a set of input Vals that are used for
+    // concretization. Here we check which inputs, if any, correspond to any of
+    // those Vals. These will be the inputs that are explicitly used in the
+    // cache Id for KernelArgumentHolder.
+    info_.input_affects_concretization_.resize(fusion->inputs().size(), false);
+    auto dyn_vals = info_.getRootDynamicVals();
+    for (const auto i : c10::irange(fusion->inputs().size())) {
+      auto input = fusion->inputs().at(i);
+      info_.input_affects_concretization_[i] =
+          input->isA<TensorView>() || dyn_vals.find(input) != dyn_vals.end();
```

Comment:
Why any input is considered affecting concretization?

Reply:
It's only considered to affect concretization if it's in `dyn_vals` or it's a `TensorView`, in which case the value `input_affects_concretization_[i]` is ignored and the extents/contiguity/dtype/etc are used in the ID string.

Reply:
Sorry, I meant to ask why any tensor is considered affecting concretization. I think my confusion was because of the name. All tensors are used for the cache lookup, but not necessarily for fusion concretization. Wouldn't it make more sense to use a different name that would indicate these are used as a key for caching?

Reply:
Ah yes. I agree it's a weird name, and tensorviews are pretty much ignored, so I could use an `unordered_set`. However, since this is the fastest path I figured checking a vector of bools would be faster than searching for each scalar. That concern is probably secondary to clarity though, so maybe it'd be better to switch to `unordered_set<Val*> concretization_affecting_scalars` or some similar name.

Reply:
Renamed to `record_scalar` which is more explicit for `lookupId`. I didn't switch to `unordered_set<Val*>` since `lookupId` works with a vector of `c10::IValue`s, so we'd need to zip with the fusion inputs to figure out which vals to record anyway. We could instead record their positions as ints, but I figured that is just about as complex as creating a bool vector as we do now.

Reply:
`info_.input_affects_concretization_` still sounds a little confusing me. Again, this is primarily for cache lookup, not fusion concretization. A tensor input may have `true` here, but it doesn't mean it affects concretization.

Reply:
When passed to `lookupId`, you're right that these are just the scalars that should be recorded (or the TensorViews). The entries for each input position are `true` iff either the input is a tensor or it appears in some scalar expression that we check during concretization (i.e. something in `leaf_dynamic_vals_`), so my intention was that inside `DynamicTransformInitialInfo` we actually _are_ tracking just the inputs that affect concretization.

Reply:
> we actually are tracking just the inputs that affect concretization.

Is that true for tensor inputs?

Reply:
Touche. It's ignored in `lookupId` anyway, but for consistency we can probably just set `TensorView` input entries to `false` for clarity.

Reply:
Well, some input tensors do affect concretization.

Reply:
OK, I'll check input TensorViews. If their extents affect concretization they will have a true value as well; otherwise false.

Reply:
IIUC, the only information we need is which scalar inputs are necessary for cache lookup. We could just have `unordered_set<int64_t> scalar_input_affect_concretization` to hold the indices of the scalar inputs in the input vector.

Reply:
Switched to `unordered_set`. As you said I think this is the clearest approach, and in many use cases the set will be empty so the check is cheap.